### PR TITLE
Annotation flag to control whether NMA certs are mounted

### PR DESCRIPTION
**Pull request by jizhuoyu**
Added a flag in annotations to control whether the NMA certs are mounted in a secret volume.
Depending on the value of the flag, the volume / volume mount are either created or not; the environment variables also vary depending on the value of the flag.

### CODE REVIEW

File: pkg/builder/builder.go [hunk 1]
```diff
@@ -575,19 +586,24 @@ func makeServerContainer(vdb *vapi.VerticaDB, sc *vapi.Subcluster) corev1.Contai
...
+				{Name: NMARootCAEnv, Value: fmt.Sprintf("%s/%s", paths.NMACertsRoot, paths.HTTPServerCACrtName)},
+				{Name: NMACertEnv, Value: fmt.Sprintf("%s/%s", paths.NMACertsRoot, corev1.TLSCertKey)},
+				{Name: NMAKeyEnv, Value: fmt.Sprintf("%s/%s", paths.NMACertsRoot, corev1.TLSPrivateKeyKey)},
+			}...)
+		} else {
+			envVars = append(envVars, []corev1.EnvVar{
+				// New model is for the NMA to read the secrets directly from k8s.
+				// We provide the secret namespace and name for this reason. Once
+				// implemented we no longer need to provide the above environment
+				// variables.
```
Comment (spilchen) [inline pkg/builder/builder.go:602]: We should update the comment.
```suggestion
				// The NMA will read the secrets directly from k8s. We provide the
				// secret namespace and name for this reason.
```

File: pkg/builder/builder.go [hunk 2]
```diff
@@ -575,19 +586,24 @@ func makeServerContainer(vdb *vapi.VerticaDB, sc *vapi.Subcluster) corev1.Contai
...
-			// We provide the secret namespace and name for this reason. Once
-			// implemented we no longer need to provide the above environment
-			// variables.
-			{Name: NMASecretNamespaceEnv, Value: vdb.ObjectMeta.Namespace},
-			{Name: NMASecretNameEnv, Value: vdb.Spec.NmaTLSSecret},
-		}...)
+		if vmeta.UseNMACertsMount(vdb.Annotations) {
+			envVars = append(envVars, []corev1.EnvVar{
+				// Old model is to provide the path to each of the certs that are
+				// mounted in the container.
```
Comment (spilchen) [inline pkg/builder/builder.go:592]: Suggest an update to the comment. I don't think we need to talk about it in terms of old/new model anymore:
```suggestion
				// Provide the path to each of the certs that are mounted in the container.
```

File: pkg/builder/builder.go [hunk 3]
```diff
@@ -286,8 +293,12 @@ func buildVolumes(vdb *vapi.VerticaDB) []corev1.Volume {
 	if vdb.GetSSHSecretName() != "" {
 		vols = append(vols, buildSSHVolume(vdb))
 	}
-	if vdb.Spec.NmaTLSSecret != "" {
-		vols = append(vols, buildHTTPServerSecretVolume(vdb))
+	if vmeta.UseVClusterOps(vdb.Annotations) {
+		if vmeta.UseNMACertsMount(vdb.Annotations) {
```
Comment (spilchen) [inline pkg/builder/builder.go:297]: Could we combine the lines?
```suggestion
	if vmeta.UseVClusterOps(vdb.Annotations) && vmeta.UseNMACertsMount(vdb.Annotations) {
```

File: pkg/builder/builder.go [hunk 4]
```diff
@@ -184,8 +187,12 @@ func buildVolumeMounts(vdb *vapi.VerticaDB) []corev1.VolumeMount {
 		volMnts = append(volMnts, buildSSHVolumeMounts()...)
 	}
 
-	if vdb.Spec.NmaTLSSecret != "" {
-		volMnts = append(volMnts, buildHTTPServerVolumeMount()...)
+	if vmeta.UseVClusterOps(vdb.Annotations) {
+		if vmeta.UseNMACertsMount(vdb.Annotations) {
```
Comment (spilchen) [inline pkg/builder/builder.go:191]: Could we combine the lines?
```suggestion
	if vmeta.UseVClusterOps(vdb.Annotations) && vmeta.UseNMACertsMount(vdb.Annotations) {
```

File: pkg/controllers/vdb/httpservercertgen_reconciler.go [hunk 5]
```diff
@@ -59,15 +59,15 @@ func (h *HTTPServerCertGenReconciler) Reconcile(ctx context.Context, _ *ctrl.Req
...
 	// If the secret name is set, check that it exists. As a convenience we will
 	// regenerate the secret using the same name.
-	if h.Vdb.Spec.NmaTLSSecret != "" {
-		nm := names.GenNamespacedName(h.Vdb, h.Vdb.Spec.NmaTLSSecret)
+	if h.Vdb.Spec.NMATLSSecret != "" {
+		nm := names.GenNamespacedName(h.Vdb, h.Vdb.Spec.NMATLSSecret)
 		secret := corev1.Secret{}
 		err := h.VRec.Client.Get(ctx, nm, &secret)
 		if errors.IsNotFound(err) {
 			h.Log.Info("httpServerTLSSecret is set but doesn't exist. Will recreate the secret.", "name", nm)
```
Comment (spilchen) [inline pkg/controllers/vdb/httpservercertgen_reconciler.go:67]: Can we update the message to indicate the name of the secret?
```suggestion
			h.Log.Info("nmaTLSSecret is set but doesn't exist. Will recreate the secret.", "name", nm)
```

File: pkg/controllers/vdb/httpservercertgen_reconciler.go [hunk 6]
```diff
@@ -59,15 +59,15 @@ func (h *HTTPServerCertGenReconciler) Reconcile(ctx context.Context, _ *ctrl.Req
...
 		} else if err != nil {
 			return ctrl.Result{},
-				fmt.Errorf("failed while attempting to reade the tls secret %s: %w", h.Vdb.Spec.NmaTLSSecret, err)
+				fmt.Errorf("failed while attempting to reade the tls secret %s: %w", h.Vdb.Spec.NMATLSSecret, err)
```
Comment (spilchen) [inline pkg/controllers/vdb/httpservercertgen_reconciler.go:70]: Its not your fault, but could you correct the spelling?
```suggestion
				fmt.Errorf("failed while attempting to read the tls secret %s: %w", h.Vdb.Spec.NMATLSSecret, err)
```

File: tests/e2e-leg-6/nma-certs-mount/20-rbac.yaml
Comment (spilchen) [inline tests/e2e-leg-6/nma-certs-mount/20-rbac.yaml:1]: I don't think you need step 20. Can you try without it?

File: tests/e2e-leg-6/nma-certs-mount/README.txt
Comment (spilchen) [inline tests/e2e-leg-6/nma-certs-mount/README.txt:1]: Lets remove this file.

File: tests/e2e-leg-6/nma-certs-mount/20-rbac.yaml
Comment (jizhuoyu) [reply tests/e2e-leg-6/nma-certs-mount/20-rbac.yaml:1]: done

File: tests/e2e-leg-6/nma-certs-mount/README.txt
Comment (jizhuoyu) [reply tests/e2e-leg-6/nma-certs-mount/README.txt:1]: done

**Review by spilchen — APPROVED** [review]

**Pull request merged**